velox/row/UnsafeRowDeserializers.cpp

```diff
@@ -0,0 +1,279 @@
+      } else {
+        memcpy(rawBuffer, valueSrcPtr, length);
+        column->setNoCopy(row, StringView(rawBuffer, length));
+        rawBuffer += length;
+      }
+    } else {
+      column->setNull(row, true);
+    }
+  }
+  return column;
+}
+
+template <>
+VectorPtr createFlatVectorFast<TypeKind::VARBINARY>(
+    const TypePtr& type,
+    column_index_t columnIdx,
+    vector_size_t numRows,
+    int64_t fieldOffset,
+    const std::vector<int64_t>& offsets,
+    const uint8_t* memoryAddress,
+    memory::MemoryPool* pool) {
+  return createFlatVectorFast<TypeKind::VARCHAR>(
+      type, columnIdx, numRows, fieldOffset, offsets, memoryAddress, pool);
+}
+
+VectorPtr createUnknownFlatVector(
+    vector_size_t numRows,
+    memory::MemoryPool* pool) {
+  const auto nulls = allocateNulls(numRows, pool, bits::kNull);
+  return std::make_shared<FlatVector<UnknownValue>>(
+      pool,
+      UNKNOWN(),
+      nulls,
+      numRows,
+      nullptr, // values
+      std::vector<BufferPtr>{}); // stringBuffers
+}
+
+bool fastSupported(const RowTypePtr& type) {
+  for (auto i = 0; i < type->size(); i++) {
+    const auto kind = type->childAt(i)->kind();
+    switch (kind) {
+      case TypeKind::ARRAY:
+      case TypeKind::MAP:
+      case TypeKind::ROW:
+        return false;
+      default:
+        break;
+    }
+  }
+  return true;
+}
+
+VectorPtr deserializeFast(
+    const uint8_t* memoryAddress,
+    const RowTypePtr& type,
+    const std::vector<int64_t>& offsets,
+    memory::MemoryPool* pool) {
+  const auto numFields = type->size();
+  const int64_t nullBitsetWidthInBytes = UnsafeRow::getNullLength(numFields);
+  std::vector<VectorPtr> columns(numFields);
+  const vector_size_t numRows = offsets.size();
+  for (auto i = 0; i < numFields; i++) {
+    const auto fieldOffset = getFieldOffset(nullBitsetWidthInBytes, i);
+    const auto& colType = type->childAt(i);
+    if (colType->kind() == TypeKind::UNKNOWN) {
+      columns[i] = createUnknownFlatVector(numRows, pool);
+    } else {
+      columns[i] = VELOX_DYNAMIC_SCALAR_TYPE_DISPATCH(
+          createFlatVectorFast,
+          colType->kind(),
+          colType,
+          i,
+          numRows,
+          fieldOffset,
+          offsets,
+          memoryAddress,
+          pool);
+    }
+  }
+
+  return std::make_shared<RowVector>(
+      pool, type, BufferPtr(nullptr), numRows, std::move(columns));
+}
+} // namespace
+
+// static
+VectorPtr UnsafeRowDeserializer::deserialize(
+    const uint8_t* memoryAddress,
+    const RowTypePtr& type,
+    const std::vector<int64_t>& offsets,
+    memory::MemoryPool* pool) {
+  if (fastSupported(type)) {
+    return deserializeFast(memoryAddress, type, offsets, pool);
+  } else {
+    std::vector<std::optional<std::string_view>> data;
+    const vector_size_t numRows = offsets.size();
+    for (auto i = 0; i < numRows; i++) {
+      const auto length =
+          (i == numRows - 1 ? offsets[i] : offsets[i + 1] - offsets[i]);
```

Comment:
if (i == numRows - 1), the length is 'offsets[i]'. Is it correct?

Reply:
Thanks for catching it, the offsets size should be numRows + 1, then we can get the correct last row length.